operator/docs/enhancements/zone_aware_replication.md

```diff
@@ -0,0 +1,460 @@
+
+### Cluster uses multiple topology keys
+
+PodTopologyCOnstraint in Kubernetes is able to handle multiple topology keys and can schedule the pods as expected. But, there is no documentation on how Cortex/Loki handles multiple topology keys. Until further information is found, a simple proposal is to contatenate the values of the different topology keys and create the $ZONE variable for the loki configuration. For eg:
+
+```
+topology.kubernetes.io/zone: "zone-a"
+kubernetes.io/hostname: "ip-172-20-114-199.ec2.internal"
+Concatenate the values
+ZONE: "zone-a-ip-172-20-114-199.ec2.internal"
+```
+
+### Open Questions
+
+- Both these methods first modify the pod to add the annotation[topology.kubernetes.io/zone: zone-a] in the pod and then modify the container to add a new env var or a volume which picks the zone value from the pod annotation via the downwardapi
+- In the Webhook method, is it safe to assume both the pod modification, followed by the container modification can be handled within the webhook?
+- In the operator method, there is going to be two merge patches of the object, one for the pod to add the annotation, and then the container to add the env variable.
+- Would it be better to handle the multiple modifications in an init container. If yes, Is that worth the overhead of  maintainence of an init container?
+- If the operator can handle this without init container isnt that better than a webhook?
+
+## Implementation History
+
+## Drawbacks
+
+### 1. Read and Write Path
+
+As suggested in the `Design` section there is no separate enabling of the Read and the Write path in the initial pass of the feature implementation. The benefits of enabling this separately is that, during live migrations we can first turn on write path zone awareness, leave it enabled for the max query lookback period, and then enable read path zone awareness. This means that at the time we turn on the read path zone awareness all ingesters that should have data for a stream will have all the data for that stream. If you enable the read path zone awareness earlier, then potentially data would be missing in queries. At this point the loki-operator deploys all the components in an all or nothing fashion. So the users should expect missing queries in the data for the max query lookback period.To avoid missing the queries, the operator needs to update the deployment process to two stages - first the components in the write path are deployed, wait the max query lookback period(currently 30s), followed by the deployment of the components in the read path.
+
+### 2. Topology Key limitations
+
+By just using the Topologykey to deploy the replicas in the different zones, we only ensure that 2 pods of different zones are not in the same node. But we cant control 2 replica pods in the same zone getting deployed on the same node. To prevent this we might want to let the user input zone details that can be used as a NodeSelector
```

Comment:
```suggestion
By just using the topology-key to deploy the replicas in the different zones, we only ensure that 2 pods of different zones are not in the same node. But we can't control 2 replica pods in the same zone getting deployed on the same node. To prevent this we might want to let the user input zone details that can be used as a NodeSelector.
```